pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,336 @@
+	policySnapshotList := &fleetv1beta1.ClusterSchedulingPolicySnapshotList{}
+	labelSelector := labels.SelectorFromSet(labels.Set{
+		fleetv1beta1.CRPTrackingLabel:      crp.Name,
+		fleetv1beta1.IsLatestSnapshotLabel: strconv.FormatBool(true),
+	})
+	listOptions := &client.ListOptions{LabelSelector: labelSelector}
+	// The scheduler lists with a cached client; this does not have any consistency concern as sources
+	// will always trigger the scheduler when a new policy snapshot is created.
+	if err := s.client.List(ctx, policySnapshotList, listOptions); err != nil {
+		klog.ErrorS(err, "Failed to list policy snapshots of a cluster resource placement", "clusterResourcePlacement", crpRef)
+		return nil, controller.NewAPIServerError(true, err)
+	}
+
+	switch {
+	case len(policySnapshotList.Items) == 0:
+		// There is no latest policy snapshot associated with the CRP; it could happen when
+		// * the CRP is newly created; or
+		// * due to an unexpected situation, the sequence of policy snapshots is in an inconsistent state.
+		//
+		// Either way, it is out of the scheduler's scope to handle such a case; the scheduler will
+		// be triggered again if the situation is corrected.
+		err := fmt.Errorf("no latest policy snapshot associated with cluster resource placement")
+		klog.ErrorS(err, "Failed to find the latest policy snapshot", "clusterResourcePlacement", crpRef)
+		return nil, err
+	case len(policySnapshotList.Items) > 1:
+		// There are multiple active policy snapshots associated with the CRP; normally this
+		// will never happen, as only one policy snapshot can be active in the sequence.
+		//
+		// Similarly, it is out of the scheduler's scope to handle such a case; the scheduler will
+		// report this unexpected occurrence but does not register it as a scheduler-side error.
+		// If (and when) the situation is corrected, the scheduler will be triggered again.
+		err := fmt.Errorf("too many active policy snapshots: %d, want 1", len(policySnapshotList.Items))
+		klog.ErrorS(err, "There are multiple latest policy snapshots associated with cluster resource placement", "clusterResourcePlacement", crpRef)
+		return nil, err
+	default:
+		// Found the one and only active policy snapshot.
+		return &policySnapshotList.Items[0], nil
+	}
+}
+
+// addSchedulerCleanupFinalizer adds the scheduler cleanup finalizer to a CRP (if it does not
+// have it yet).
+func (s *Scheduler) addSchedulerCleanUpFinalizer(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement) error {
+	// Add the finalizer only if the CRP does not have one yet.
+	if !controllerutil.ContainsFinalizer(crp, fleetv1beta1.SchedulerCRPCleanupFinalizer) {
+		controllerutil.AddFinalizer(crp, fleetv1beta1.SchedulerCRPCleanupFinalizer)
+
+		if err := s.client.Update(ctx, crp); err != nil {
+			klog.ErrorS(err, "Failed to update cluster resource placement", "clusterResourcePlacement", klog.KObj(crp))
+			return controller.NewAPIServerError(false, err)
```

Comment:
we now have NewUpdateIgnoreConflictError

Reply:
Done. It seems that the PR was sent before the method is added